nixos/nextcloud: Update recommended nginx settings

bachp opened the pull request:
###### Motivation for this change

Update default configuration to match the latest recommendations from Nextcloud: https://docs.nextcloud.com/server/15/admin_manual/installation/nginx.html

To get rid of all the warnings the `imagick` PHP module is enabled too.

Also update to PHP 7.3 as this is officially supported by Nextcloud 15: https://docs.nextcloud.com/server/15/admin_manual/installation/system_requirements.html

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bachp committed:
nixos/nextcloud: update recommended nginx settings

This updates the configuration to the recommendations in
https://docs.nextcloud.com/server/15/admin_manual/installation/nginx.html

bachp committed:
nixos/nextcloud: use PHP 7.3 instead of 7.1 by default

dasJ commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -360,7 +361,7 @@ in {
       services.phpfpm = {
         phpOptions = phpOptionsExtensions;
-        phpPackage = pkgs.php71;
+        phpPackage = pkgs.php73;
```
Well, if this is hard coded, why is `phpPackages` modifyable? Either both or nothing, everything else just doesn't make any sense

bachp replied:
I agree. I have no idea why `phpPackages` is configurable. Maybe it would be best to just remove the config option.

dasJ replied:
Yes

bachp replied:
I removed the `phpPackages` configuration option.

dasJ commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -469,10 +471,12 @@ in {
               error_page 404 /core/templates/404.php;
               client_max_body_size ${cfg.maxUploadSize};
               fastcgi_buffers 64 4K;
+              fastcgi_hide_header X-Powered-By;
```
Good change, should probably be added to the general nginx configuration at some point IMO

bachp replied:
@dasJ Do you mean global to the nginx module or the global nginx module of nextcloud? Adding it to the global nginx module is probably not a good idea, as there is no way to undo it if somebody wants to not hide the header.

dasJ replied:
@bachp Yes, I mean the global nginx module. It was just a thought, nothing to be added to this PR ;)

aanderse replied:
I did something similar for Apache in https://github.com/NixOS/nixpkgs/commit/1bec75301bd9d15c2ee0be4c587bc652f65df582

I think it's a great idea for another PR 👍

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -360,7 +361,7 @@ in {
 
       services.phpfpm = {
         phpOptions = phpOptionsExtensions;
```
If you don't mind would you consider moving this line *into* the pool? See https://github.com/NixOS/nixpkgs/pull/55742 for details.

bachp replied:
@aanderse If I understand correctly both `phpOptions` and `phpPackage` can be moved to the pool right?

bachp replied:
Done, both moved to pool

aanderse commented:
Yes! Which is great for reproducible systems!

bachp committed:
nixos/nextcloud: move phpPackage and phpOptions into pool

This allows to have a php configuration for nextcloud that is independent
of the global configuration.

bachp committed:
nixos/nextcloud: don't make phpPackages configurable

It needs to match the version in phpfm which is hard coded.
So there is no point in being able to change it.

fpletz commented:
@GrahamcOfBorg test nextcloud.basic nextcloud.with-mysql-and-memcached nextcloud.with-postgresql-and-redis
